### ISSUE TITLE
Crash in libpng loading the icon from NFL Street 2 Unleashed - repro with just savefile

### ISSUE THREAD

**Issue by hrydgard** [hrydgard/ppsspp#14628]
Thanks ocn for narrowing down the game!

(Easy repro without the game by just downloading the save from https://gamefaqs.gamespot.com/psp/925748-nfl-street-2-unleashed/saves , installing it in PSP/SAVEDATA and entering Settings/Tools/Savedata Manager.)

When PPSSPP tries to load the icon of NFL Street 2 Unleashed, either from the ISO in the game browser or from the save data, it crashes deep in libpng. Specifically, on [line 4000 exactly in pngread.c](https://github.com/hrydgard/ppsspp/blob/master/ext/libpng17/pngread.c#L4000):

```
      /* This is actually an internal error. */
      affirm(out_format == format /* else unimplemented transformations */);
```

out_format is 2 but format is 3. I find the code around here completely unreadable, but at least that flags field has the following definitions:

```
#define PNG_FORMAT_FLAG_ALPHA    0x01U /* format with an alpha channel */
#define PNG_FORMAT_FLAG_COLOR    0x02U /* color format: otherwise grayscale */
```

So I guess alpha gets lost somewhere and it gets confused?

I really thought libpng was better than this :P 

Attached the png here, zipped to avoid any browser bugs: 

[ICON0.zip](https://github.com/hrydgard/ppsspp/files/6842116/ICON0.zip)

**Comment by hrydgard** [comment]
I hacked away the assert for now, which creates a squished image instead of a crash.

This does seem to be a bug in libpng.

**Comment by Panderner** [comment]
@hrydgard is this libpng bug?

**Comment by hrydgard** [comment]
Yeah, it appears to be.

**Comment by Panderner** [comment]
Note: The last build work is v1.11.3-672-gf1bf596f5
The issue started in v1.11.3-677-g5fa1af6ef

**Comment by Panderner** [comment]
Here's a screenshot with proper NFL Street 2 Unleashed icon in v1.11.3-672-gf1bf596f5 and older:
![Screenshot_2021-07-28-23-20-50-595_org ppsspp ppsspp](https://user-images.githubusercontent.com/37503397/127349511-565ddcc4-b414-48ad-9ef8-d17aeb681999.jpg)

**Comment by hrydgard** [comment]
Yeah, @unknownbrackets updated libpng to a new version then in #14469, which seems to have this bug. Maybe we should just go back.

**Comment by unknownbrackets** [comment]
Yeah, we may want to revert - potentially only back to the version before this bug was introduced, but that sounds annoying to hunt down.

-[Unknown]

**Comment by unknownbrackets** [comment]
Basically, this happens after `check_tRNS_for_alpha` determines that there's alpha data, but it's all 255 (i.e. it's actually fully opaque.)  As an optimization, it ignores the alpha data since it's all known at this point.  `init_caching` then skips setting the format alpha flag.

After this, `png_init_alpha` runs, but doesn't see any need to do anything because the alpha data was marked as zero.  There's also no need to set any background or gamma transform.

Then we get to `png_init_expand`.  This sets itself up to expand the palette, without alpha.

Within `png_image_read_direct`, `png_image_format(png_ptr)` calculates the image format as RGBA (despite the fact that we've "optimized out" the A and are only outputting RGB.)  This is because it runs before the optimization that sets `num_trans` to zero noted above (`check_tRNS_for_alpha()`) which is triggered by `png_read_update_info(png_ptr, info_ptr);`.  Because of this, it never adds the `png_set_add_alpha()` transform.

I'm going to disable the optimization.  From what I can tell, everything will work fine without it, it'll just read the tRNS transparency values when assembling the image data.  It seems like a more worthwhile optimization if you're looking for RGB and probably was added with that in mind.  libpng1.6 doesn't seem to have this optimization.

-[Unknown]

**Comment by jbowler** [comment]
Yes, that's a bug.  I draw your attention to this tell-tale comment right at the top:

>Last changed in libpng 1.7.0 [(PENDING RELEASE)]

"PENDING RELEASE" means you are using unreleased code.  Stick to 1.6, 1.7 is a complete rewrite and it was never finished; Glenn numbers for the *NEXT* release, so no very of 1.7 has ever been released; indeed the current tip of the repo is "1.7.0beta90".

The changes in this version are so extreme (it really is a libpng 2.0) that I doubt it will ever be released.

**Comment by unknownbrackets** [comment]
We were previously on an older 1.7 beta (I think there was some ARM bug that motivated this, perhaps it is now fixed in 1.6), and we updated to the latest 1.7 beta because, well, being on an older beta is worse than being on a newer beta, *probably*.

If it helps, PPSSPP has been using libpng 1.7 since 2014 (39d48a7e67db693ec7c34a1c0121546c521bfebe - 1.7.0beta35) and more recently updated to the latest beta this year (8eb773888f8aa39b9f15902fb2bfadbc8d9f4656 - 1.7.0beta90.)  The previous beta, at least, has worked for some hundred million users or more over that time, and so far this is the only bug (in our usage) in the new beta.

Lots of users of PNGs and lots of PNGs out there, to be sure, but at least some tiny fraction of the world has already been exposed to this new version relatively unscathed.

-[Unknown]

**Comment by hrydgard** [comment]
Huh, I was under the impression that 1.7.0 was solid. Numbering for the next release is a strange convention that creates the wrong expectations, IMHO.
